# Python Programming Fundamentals

## Python Built-in String Methods 

A string is a sequence of characters enclosed in quotation marks.

In [1]:
# str??

In [2]:
print(dir(str))

['__add__', '__class__', '__contains__', '__delattr__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__getnewargs__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__iter__', '__le__', '__len__', '__lt__', '__mod__', '__mul__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__rmod__', '__rmul__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', 'capitalize', 'casefold', 'center', 'count', 'encode', 'endswith', 'expandtabs', 'find', 'format', 'format_map', 'index', 'isalnum', 'isalpha', 'isascii', 'isdecimal', 'isdigit', 'isidentifier', 'islower', 'isnumeric', 'isprintable', 'isspace', 'istitle', 'isupper', 'join', 'ljust', 'lower', 'lstrip', 'maketrans', 'partition', 'replace', 'rfind', 'rindex', 'rjust', 'rpartition', 'rsplit', 'rstrip', 'split', 'splitlines', 'startswith', 'strip', 'swapcase', 'title', 'translate', 'upper', 'zfill']


#### String Find Method

In [3]:
str.find?

Docstring:
S.find(sub[, start[, end]]) -> int

Return the lowest index in S where substring sub is found,
such that sub is contained within S[start:end].  Optional
arguments start and end are interpreted as in slice notation.

Return -1 on failure.
Type:      method_descriptor


The find() method returns the index of first occurrence of the substring (if found). If not found, it returns -1.

<b>The syntax of find() method is:</b>

<code>str.find(sub[, start[, end]])</code>

<b>String find() Parameters()</b>

- The find() method takes maximum of three parameters:

    - <b>sub</b> - It's the substring to be searched in the <code>str</code> string.
    - <b>start</b> and <b>end</b> (optional) - substring is searched within <code>str[start:end]</code>

<b>Return Value from find()</b>

- The find() method returns an integer value.

    - If substring exists inside the string, it returns the index of first occurence of the substring.
    - If substring doesn't exist inside the string, it returns -1.

<b>Example 1: find() With No start and end Argument</b>

In [4]:
s1 = 'rizwan.datahub@gmail.com' 

In [5]:
print(s1.find('gmail'))

15


In [6]:
print(s1.find('com'))

21


In [7]:
print(s1.find('outlook'))

-1


In [8]:
print(s1.find('outlook') == -1)

True


In [9]:
print(s1.find(' '))

-1


<b>Example 2: find() With start and end Arguments</b>

In [10]:
s1

'rizwan.datahub@gmail.com'

In [11]:
print(s1.find('@'))

14


In [12]:
print(s1.find('@', 10))

14


In [13]:
print(s1.find('@', 16))

-1


In [14]:
print(s1.find('@', 10, 15))

14


In [15]:
print(s1.find('@', 15, 20))

-1


#### Solution - String Parsing Problem 

In [16]:
m = 'From rizwan.datahub@gmail.com Wed February 24 20:05:37 2020'

Expected Output: <code>gmail.com</code>

In [17]:
print(m[m.find('@')+1 : m.find(' ', m.find('@'))])  

gmail.com


#### String Format Method

In [18]:
str.format?

Docstring:
S.format(*args, **kwargs) -> str

Return a formatted version of S, using substitutions from args and kwargs.
The substitutions are identified by braces ('{' and '}').
Type:      method_descriptor


The string format() method formats the given string into a nicer output in Python.

Link1: https://docs.python.org/3/library/stdtypes.html#str.format

Link2: https://docs.python.org/3/library/string.html#formatstrings

<b>The syntax of format() method is:</b>

<code>template_string.format(pos0, pos1, ..., key0=val0, key1=val1, ...)</code>

Here, <code>pos0, pos1,...</code> are positional arguments and, <code>key0, key1,...</code> are keyword arguments with values <code>val0, val1,...</code> respectively.

And, <code>template_string</code> is a mixture of format codes with placeholders for the arguments.

<b>String format() Parameters</b>

- format() method takes any number of parameters. But, is divided into two types of parameters:

    - <b>Positional parameters</b> - list of parameters that can be accessed with index of parameter inside curly braces {index}
    - <b>Keyword parameters</b> - list of parameters of type key=value, that can be accessed with key of parameter inside curly braces {key}

<b>Return value from String format()</b>

The format() method returns the formatted string.

<b>How String format() works?</b>

The format() reads the type of arguments passed to it and formats it according to the format codes defined in the string.

<u>For positional arguments</u>

In [19]:
"Hi {0}, your current amount is {1:9.2f}".format("Rizwan", 50000.989)

# "Rizwan"    - Argument 0 
# 50000.989   - Argument 1
# 'Hi Rizwan, your current amount is 050000.99'

'Hi Rizwan, your current amount is  50000.99'

Here, Argument 0 is a string "Rizwan" and Argument 1 is a floating number 50000.989.

<b>Note:</b> Argument list starts from 0 in Python.

The string <code>"Hi {0}, your current amount is {1:9.2f}"</code> is the template string. This contains the format codes for formatting.

The curly braces are just placeholders for the arguments to be placed. In the above example, <code>{0}</code> is placeholder for <code>"Rizwan"</code> and <code>{1:9.2f}</code> is placeholder for <code>50000.989</code>.

Since the template string references format() arguments as {0} and {1}, the arguments are positional arguments. They both can also be referenced without the numbers as {} and Python internally converts them to numbers.

Internally,

- Since <code>"Rizwan"</code> is the 0<sup>th</sup> argument, it is placed in place of {0}. Since, {0} doesn't contain any other format codes, it doesn't perform any other operations.
- However, it is not the case for 1<sup>st</sup> argument <code>50000.989</code>. Here, <code>{1:9.2f}</code> places <code>50000.989</code> in its place and performs the operation 9.2f.
- f specifies the format is dealing with a float number. If not correctly specified, it will give out an error.
- The part before the "." (9) specifies the minimum width/padding the number (50000.989) can take. In this case, 50000.989 is allotted a minimum of 9 places including the ".".
- If no alignment option is specified, it is aligned to the right of the remaining spaces. (For strings, it is aligned to the left.)
- The part after the "." (2) truncates the decimal part (989) upto the given number. In this case, 989 is truncated after 2 places.
- Remaining numbers (9) is rounded off outputting 99.

<u>For keyword arguments</u>

In [20]:
"Hi {name}, your current amount is {amt:9.2f}".format(name="Rizwan", amt=50000.989)

'Hi Rizwan, your current amount is  50000.99'

I've used the same example from above to show the difference between keyword and positional arguments.

Here, instead of just the parameters, we've used a key-value for the parameters. Namely, <code>name="Rizwan"</code> and <code>amt=50000.989</code>.

Since, these parameters are referenced by their keys as <code>{name}<code> and <code>{amt:9.2f}</code>, they are known as keyword or named arguments.

Internally,

- The placeholder <code>{name}</code> is replaced by the value of name - "Rizwan". Since, it doesn't contain any other format codes, "Rizwan" is placed.
- For the argument amt=50000.989, the placeholder <code>{amt:9.2f}</code> is replaced by the value 50000.989. But before replacing it, like previous example, it performs 9.2f operation on it.
- This outputs  50000.99. The decimal part is truncated after 2 places and remaining digits are rounded off. Likewise, the total width is assigned 9 leaving one spaces to the left.

<b><u>Basic formatting with format()</u></b>

The format() method allows the use of simple placeholders for formatting.

<b>Example 1: Basic formatting for default, positional and keyword arguments</b>

In [21]:
# default arguments
print("Hi {}, your current amount is {}.".format("Rizwan", 50000.989))

Hi Rizwan, your current amount is 50000.989.


In [22]:
# positional arguments
print("Hi {0}, your current amount is {1}.".format("Rizwan", 50000.989))

Hi Rizwan, your current amount is 50000.989.


In [23]:
# keyword arguments
print("Hi {name}, your current amount is {amt}.".format(name="Rizwan", amt=50000.989))

Hi Rizwan, your current amount is 50000.989.


In [24]:
# mixed arguments
print("Hi {0}, your current amount is {amt}.".format("Rizwan", amt=50000.989))

Hi Rizwan, your current amount is 50000.989.


<b>Note</b>: In case of mixed arguments, keyword arguments has to always follow positional arguments.

<b>Numbers formatting with format()</b>

You can format numbers using the format specifier given below:

<code>Type - Meaning
d - Decimal integer
c - Corresponding Unicode character
b - Binary format
o - Octal format
x - Hexadecimal format (lower case)
X - Hexadecimal format (upper case)
n - Same as 'd'. Except it uses current locale setting for number separator
e - Exponential notation. (lowercase e)
E - Exponential notation (uppercase E)
f - Displays fixed point number (Default: 6)
F - Same as 'f'. Except displays 'inf' as 'INF' and 'nan' as 'NAN'
g - General format. Rounds number to p significant digits. (Default precision: 6)
G - Same as 'g'. Except switches to 'E' if the number is large.
% - Percentage. Multiples by 100 and puts % at the end.</code>

<b>Example 2: Simple number formatting</b>

In [25]:
# integer arguments
print("The number is - {0:d}".format(123))

The number is - 123


In [26]:
# float arguments
print("The float number is - {0:f}".format(123.4567898))

The float number is - 123.456790


In [27]:
# octal, binary and hexadecimal format
print("bin - {0:b}, oct - {0:o}, hex - {0:X}".format(15))

bin - 1111, oct - 17, hex - F


<b>Example 3: Number formatting with padding for int and floats</b>

In [28]:
# integer numbers with minimum width
print("{0:5d}".format(12))

   12


In [29]:
# width doesn't work for numbers longer than padding
print("{:2d}".format(1234))

1234


In [30]:
# padding for float numbers
print("{:8.3f}".format(12.2346))

  12.235


In [31]:
# integer numbers with minimum width filled with zeros
print("{:05d}".format(12))

00012


In [32]:
# padding for float numbers filled with zeros
print("{:08.3f}".format(12.2346))

0012.235


Here,

- In the first statement, {0:5d} takes an integer argument and assigns a minimum width of 5. Since, no alignment is specified, it is aligned to the right.
- In the second statement, you can see the width (2) is less than the number (1234), so it doesn't take any space to the left but also doesn't truncate the number.
- Unlike integers, floats has both integer and decimal parts. And, the mininum width defined to the number is for both parts as a whole including ".".
- In the third statement, {:8.3f} truncates the decimal part into 3 places rounding off the last 2 digits. And, the number, now 12.235, takes a width of 8 as a whole leaving 2 places to the left.
- If you want to fill the remaining places with zero, placing a zero before the format specifier does this. It works both for integers and floats: {:05d} and {:08.3f}.

<b>Number formatting with alignment</b>

The operators <, ^, > and = are used for alignment when assigned a certain width to the numbers.

<  Left aligned to the remaining space

^  Center aligned to the remaining space

<code>></code> Right aligned to the remaining space

=  Forces the signed (+) (-) to the leftmost position

<b>Example 4: Number formatting with left, right and center alignment</b>

In [33]:
# integer numbers with right alignment
print("{:5d}".format(12))

   12


In [34]:
# float numbers with center alignment
print("{:*^10.3f}".format(12.2346))

**12.235**


<b>Example 5: String formatting with padding and alignment</b>

In [35]:
# string padding with left alignment
print("{:8}".format("python"))

python  


In [36]:
# string padding with right alignment
print("{:>8}".format("python"))

  python


In [37]:
# string padding with center alignment
print("{:^8}".format("python"))

 python 


In [38]:
# string padding with center alignment
# and '*' padding character
print("{:*^8}".format("python"))

*python*


#### Python String Formatting

- 'Old Style' String Concatenation (+ Operator)
- 'Old Style' String Formatting (% Operator)
- <b>'New Style' String Formatting (str.format)</b>
- <b>String Interpolation / f-Strings (Python 3.6+)</b>
- Template Strings (Standard Library)

<u>'Old Style' String Concatenation (+ Operator)</u>

In [39]:
name = "ahamd"
cgpa = 3.9

In [40]:
name + str(cgpa)

'ahamd3.9'

In [41]:
name + " " + str(cgpa)

'ahamd 3.9'

In [42]:
print('My Info - Name : ' + name + ', CGPA : ' + str(cgpa))

My Info - Name : ahamd, CGPA : 3.9


<u>'Old Style' String Formatting (% Operator)</u>

https://docs.python.org/3/library/stdtypes.html#old-string-formatting

<u><b>'New Style' String Formatting (str.format)</b></u>

In [43]:
print('My Info - Name : {}, CGPA : {}'.format(cgpa, name))

My Info - Name : 3.9, CGPA : ahamd


In [44]:
print('My Info - Name : {1}, CGPA : {0}'.format(cgpa, name))

My Info - Name : ahamd, CGPA : 3.9


In [45]:
print('My Info - Name : {name}, CGPA : {cgpa}'.format(cgpa=3.9, name='ahmad'))

My Info - Name : ahmad, CGPA : 3.9


<u><b>String Interpolation / f-Strings (Python 3.6+)</u></b>

In [46]:
print(f'My Info - Name : {name}, CGPA : {cgpa}')

My Info - Name : ahamd, CGPA : 3.9


(f-string) 

PEP 3101, PEP 498, PEP 501

#### Happy Learning 😊